Landmark based facial expression recognition (#95)
opendr-eu/opendr: docs/reference/landmark-based-facial-expression-recognition.md

- **temp_path**: *str, default=''*  
  Specifies a path where the algorithm saves the checkpoints and onnx optimized model (if needed).
- **device**: *{'cpu', 'cuda'}, default='cuda'*  
  Specifies the device to be used.
- **num_workers**: *int, default=32*  
  Specifies the number of workers to be used by the data loader.
- **epochs**: *int, default=400*  
  Specifies the number of epochs the training should run for.
- **experiment_name**: *str, default='pstbln_casia'*  
  String name to attach to checkpoints.
- **device_ind**: *list, default=[0]*  
  List of GPU indices to be used if the device is 'cuda'. 
- **val_batch_size**: *int, default=256*  
  Specifies number of samples to be bundled up in a batch during evaluation. This heavily affects memory usage, adjust according to your system.
- **drop_after_epoch**: *list, default=[1000]*  
  List of epoch numbers in which the optimizer drops the learning rate.  
- **start_epoch**: *int, default=0*  
  Specifies the starting epoch number for training.  
- **dataset_name**: *str {'AFEW', 'CASIA', 'CK+'}, default='CASIA'*  
  Specifies the name of dataset that is used for training and evaluation. 

> - **block_size**: *int, default=5*  

Reviewer: here the argument name is correct.
You should also change it in the code where `blocksize` is used.

  Specifies the number of output channels (or neurons) that are added to each layer of the network at each progression iteration. 

> - **numblocks**: *int, default=100*  

Reviewer: This argument name format is not consistent with the others.
It should rather be `num_blocks` (same for `num_layers`).

- **topology**: *list, default=[]*  
  Specifies the initial topology of the network. The default is set to [], since the method gets an empty network as input and builds it progressively. 
- **layer_threshold**: *float, default=1e-4*  
  Specifies the threshold which is used by the method to identify when it should stop adding new layers.  
- **block_threshold**: *float, default=1e-4*  
  Specifies the threshold which is used by the model to identify when it should stop adding new blocks in each layer.  
  

#### `ProgressiveSpatioTemporalBLNLearner.fit`
```python
ProgressiveSpatioTemporalBLNLearner.fit(self, dataset, val_dataset, logging_path, silent, verbose,
                                        momentum, nesterov, weight_decay, monte_carlo_dropout, 
                                        mcdo_repeats, train_data_filename,
                                        train_labels_filename, val_data_filename,
                                        val_labels_filename)
```

This method is used for training the algorithm on a train dataset and validating on a val dataset.
Parameters:
- **dataset**: *object*  
Object that holds the training dataset.

> Can be of type `ExternalDataset` or a custom dataset inheriting from `DatasetIterator`.

Reviewer: It is rendered correctly, but all the lines in the fit method should be indented by 2 spaces.

- **val_dataset**: *object*
Object that holds the validation dataset. 
- **logging_path**: *str, default=''*  
Path to save TensorBoard log files and the training log files.
If set to None or '', TensorBoard logging is disabled and no log file is created. 
- **silent**: *bool, default=False*  
If set to True, disables all printing of training progress reports and other information to STDOUT.
- **verbose**: *bool, default=True***  
If set to True, enables the maximum verbosity.
- **momentum**: *float, default=0.9*  
Specifies the momentum value for optimizer. 
- **nesterov**: *bool, default=True***  

> If set to true, the optimizer uses nesterov.  

Reviewer: Here I would better explain what is `nesterov`, for example
```suggestion
  If set to true, the optimizer uses Nesterov's momentum.  
```

- **weight_decay**: *float, default=0.0001***  
Specifies the weight_decay value of the optimizer. 
- **monte_carlo_dropout**: *bool, default=True*** 

> If set to True, enables the Monte Carlo Dropout in inference

Reviewer: ```suggestion
  If set to True, enables the Monte Carlo Dropout in inference.
```

- **mcdo_repeats**: *int, default=100*** 

> Specifies the number of times that inference is repeated for Monte Carlo Dropout

Reviewer: ```suggestion
  Specifies the number of times that inference is repeated for Monte Carlo Dropout.
```

- **train_data_filename**: *str, default='train.npy'*  
Filename that contains the training data. 
This file should be contained in the dataset path provided.
Note that this is a file name, not a path.
- **train_labels_filename**: *str, default='train_labels.pkl'*  
Filename of the labels .pkl file. 
This file should be contained in the dataset path provided.
- **val_data_filename**: *str, default='val.npy'*  
Filename that contains the validation data.
This file should be contained in the dataset path provided.
Note that this is a filename, not a path.
- **val_labels_filename**: *str, default='val_labels.pkl'*  
Filename of the validation labels .pkl file.
This file should be contained in the dataset path provided.

#### `ProgressiveSpatioTemporalBLNLearner.eval`
```python

> ProgressiveSpatioTemporalBLNLearner.eval(self, val_dataset, epoch, monte_carlo_dropout, 

Reviewer: `val_loader` argument is missing

                                         mcdo_repeats, silent, verbose,
                                         val_data_filename, val_labels_filename,
                                         save_score, wrong_file, result_file, show_topk)
```

This method is used to evaluate a trained model on an evaluation dataset.
Returns a dictionary containing stats regarding evaluation.  
Parameters:

- **val_dataset**: *object*  
  Object that holds the evaluation dataset.
  Can be of type `ExternalDataset` or a custom dataset inheriting from `DatasetIterator`.
- **epoch**: *int, default=0*  
  The training epoch in which the model is evaluated. 
- **monte_carlo_dropout**: *bool, default=True*** 

>   If set to True, enables the Monte Carlo Dropout in inference

Reviewer: ```suggestion
  If set to True, enables the Monte Carlo Dropout in inference.
```

- **mcdo_repeats**: *int, default=100*** 

>   Specifies the number of times that inference is repeated for Monte Carlo Dropout

Reviewer: ```suggestion
  Specifies the number of times that inference is repeated for Monte Carlo Dropout.
```

  Filename that contains the validation data.
  This file should be contained in the dataset path provided.
  Note that this is a filename, not a path.
- **val_labels_filename**: *str, default='val_labels.pkl'*  
  Filename of the validation labels .pkl file.
  This file should be contained in the dataset path provided.
- **save_score**: *bool, default=False*  
  If set to True, it saves the classification score of all samples in differenc classes
  in a log file. Default to False.
- **wrong_file**: *str, default=None*  
  If set to True, it saves the results of wrongly classified samples. Default to False.
- **result_file**: *str, default=None*  
  If set to True, it saves the classification results of all samples. Default to False.
- **show_topk**: *list, default=[1, 5]*  
  Is set to a list of integer numbers defining the k in top-k accuracy. Default is set to [1,5].

#### `ProgressiveSpatioTemporalBLNLearner.init_model`
```python
ProgressiveSpatioTemporalBLNLearner.init_model(self)
```
This method is used to initialize the imported model and its loss function. 
  
  
#### `ProgressiveSpatioTemporalBLNLearner.network_builder`
```python

> ProgressiveSpatioTemporalBLNLearner.network_builder(self, dataset, val_dataset, monte_carlo_dropout, 

Reviewer: `logging_path` argument is missing

  This file should be contained in the dataset path provided.
  Note that this is a file name, not a path.
- **train_labels_filename**: *str, default='train_labels.pkl'*  
  Filename of the labels .pkl file. 
  This file should be contained in the dataset path provided.
- **val_data_filename**: *str, default='val.npy'*  
  Filename that contains the validation data.
  This file should be contained in the dataset path provided.
  Note that this is a filename, not a path.
- **val_labels_filename**: *str, default='val_labels.pkl'*  
  Filename of the validation labels .pkl file.
  This file should be contained in the dataset path provided.
- **verbose**: *bool, default=True***  
  Whether to print messages in the console.
  
  
#### `ProgressiveSpatioTemporalBLNLearner.infer`
```python
ProgressiveSpatioTemporalBLNLearner.infer(self, facial_landmarks_batch, monte_carlo_dropout, mcdo_repeats)
```
This method is used to perform action recognition on a sequence of landmarks. 
It returns the action category as an object of `engine.target.Category` if a proper input object `engine.data.SkeletonSequence` is given. 

Parameters:
- **facial_landmarks_batch**: *object***  

>   Object of type engine.data.SkeletonSequence.

Reviewer: ```suggestion
  Object of type `engine.data.SkeletonSequence`.
```

- **monte_carlo_dropout**: *bool, default=True*** 

>   If set to True, enables the Monte Carlo Dropout in inference

Reviewer: ```suggestion
  If set to True, enables the Monte Carlo Dropout in inference.
```

- **mcdo_repeats**: *int, default=100*** 

>   Specifies the number of times that inference is repeated for Monte Carlo Dropout

Reviewer: ```suggestion
  Specifies the number of times that inference is repeated for Monte Carlo Dropout.
```

#### `ProgressiveSpatioTemporalBLNLearner.save`
```python
ProgressiveSpatioTemporalBLNLearner.save(self, path, model_name, verbose)
```
This method is used to save a trained model.
Provided with the path "/my/path" (absolute or relative), it creates the "path" directory, if it does not already 
exist. Inside this folder, the model is saved as "model_name.pt" and the metadata file as "model_name.json". If the directory
already exists, the "model_name.pt" and "model_name.json" files are overwritten.

If [`self.optimize`](#ProgressiveSpatioTemporalBLNLearner.optimize) was run previously, it saves the optimized ONNX model in 
a similar fashion with an ".onnx" extension, by copying it from the self.temp_path it was saved previously 
during conversion.

Parameters:
- **path**: *str*  
  Path to save the model.

> - **model_name**: *str*  

Reviewer: In code the default value `''` is provided.

  The file name to be saved. 

> - **verbose**: *bool, default=False*  

Reviewer: Mismatching default value with code
```suggestion
- **verbose**: *bool, default=True*  
```

  If set to True, prints a message on success.

#### `ProgressiveSpatioTemporalBLNLearner.load`
```python
ProgressiveSpatioTemporalBLNLearner.load(self, path, model_name, verbose)
```

This method is used to load a previously saved model from its saved folder.
Loads the model from inside the directory of the path provided, using the metadata .json file included.

Parameters:
- **path**: *str*  
  Path of the model to be loaded.
- **model_name**: *str*  
  The file name to be loaded. 

> - **verbose**: *bool, default=False*  

Reviewer: Mismatching default value with code
```suggestion
- **verbose**: *bool, default=True*  
```

  If set to True, prints a message on success.

#### `ProgressiveSpatioTemporalBLNLearner.optimize`
```python
ProgressiveSpatioTemporalBLNLearner.optimize(self, do_constant_folding)
```

This method is used to optimize a trained model to ONNX format which can be then used for inference.

Parameters:
- **do_constant_folding**: *bool, default=False*  
  ONNX format optimization.
  If True, the constant-folding optimization is applied to the model during export.
  Constant-folding optimization will replace some of the operations that have all constant inputs, with pre-computed constant nodes.

#### `ProgressiveSpatioTemporalBLNLearner.download`
```python
@staticmethod
ProgressiveSpatioTemporalBLNLearner.download(self, path, mode, verbose, url)
```
Downloads files depending on mode and saves them in the path provided. It supports downloading train, validation and test dataset containing one or more landmark sequences and its labels.  

Parameters:
- **path**: *str, default=None*  

>   Local path to save the files, defaults to self.parent_dir if None.

Reviewer: ```suggestion
  Local path to save the files, defaults to `self.parent_dir` if None.
```

- **mode**: *str, default="train_data"*  
  What file to download, can be one of "pretrained", "train_data", "val_data", "test_data"

> - **verbose**: *bool, default=False*  

Reviewer: Mismatching default value with code
```suggestion
- **verbose**: *bool, default=True*  
```

```python
ProgressiveSpatioTemporalBLNLearner(self, lr, batch_size, optimizer_name, lr_schedule,
                                    checkpoint_after_iter, checkpoint_load_iter, temp_path,
                                    device, num_workers, epochs, experiment_name,
                                    device_indices, val_batch_size, drop_after_epoch,
                                    start_epoch, dataset_name, num_class, num_point,  
                                    num_person, in_channels, 
                                    block_size, num_blocks, num_layers, topology,
                                    layer_threshold, block_threshold)
```

Constructor parameters:
- **lr**: *float, default=0.1*  
  Specifies the initial learning rate to be used during training.
- **batch_size**: *int, default=128*  
  Specifies number of samples to be bundled up in a batch during training. This heavily affects memory usage, adjust according to your system.
- **optimizer_name**: *str {'sgd', 'adam'}, default='sgd'*  
  Specifies the optimizer type that should be used.
- **lr_schedule**: *str, default=' '*  
  Specifies the learning rate scheduler.

> - **checkpoint_after_iter**: *int, default=0*

Reviewer 2: ```suggestion
- **checkpoint_after_iter**: *int, default=0*  
```

Reviewer 2: Consistent formatting

  in a log file. Default to False.
- **wrong_file**: *str, default=None*  
  If set to True, it saves the results of wrongly classified samples. Default to False.
- **result_file**: *str, default=None*  
  If set to True, it saves the classification results of all samples. Default to False.
- **show_topk**: *list, default=[1, 5]*  
  Is set to a list of integer numbers defining the k in top-k accuracy. Default is set to [1,5].

#### `ProgressiveSpatioTemporalBLNLearner.init_model`
```python
ProgressiveSpatioTemporalBLNLearner.init_model(self)
```
This method is used to initialize the imported model and its loss function. 
  
  
#### `ProgressiveSpatioTemporalBLNLearner.network_builder`
```python
ProgressiveSpatioTemporalBLNLearner.network_builder(self, dataset, val_dataset, monte_carlo_dropout, 
                                                    mcdo_repeats, logging_path, train_data_filename,
                                                    train_labels_filename, val_data_filename,
                                                    val_labels_filename, verbose)
```
This method builds the network topology progressively. 

> Parameters:

Reviewer 2: ```suggestion

Parameters:
```
